mongo/mongo.go

```diff
@@ -0,0 +1,189 @@
+}
+
+// newMongoStore returns a new MongoDB session store based on given configuration.
+func newMongoStore(cfg Config) *mongoStore {
+	return &mongoStore{
+		nowFunc:    cfg.nowFunc,
+		lifetime:   cfg.Lifetime,
+		db:         cfg.db,
+		collection: cfg.Collection,
+		encoder:    cfg.Encoder,
+		decoder:    cfg.Decoder,
+	}
+}
+
+func (s mongoStore) Exist(ctx context.Context, sid string) bool {
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Err()
+	return err != mongo.ErrNoDocuments
+}
+
+func (s mongoStore) Read(ctx context.Context, sid string) (session.Session, error) {
+	var result bson.M
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Decode(&result)
+	if err == nil {
+		binary, ok := result["data"].(primitive.Binary)
+		if !ok {
+			return nil, errors.New("assert `data` key")
+		}
+
+		expiredAt, ok := result["expired_at"].(primitive.DateTime)
+		if !ok {
+			return nil, errors.New("assert `expired_at` key")
+		}
+
+		// Discard existing data if it's expired
+		if !s.nowFunc().Before(expiredAt.Time().Add(s.lifetime)) {
+			return session.NewBaseSession(sid, s.encoder), nil
+		}
+
+		data, err := s.decoder(binary.Data)
+		if err != nil {
+			return nil, errors.Wrap(err, "decode")
+		}
+
+		sess := session.NewBaseSession(sid, s.encoder)
+		sess.SetData(data)
+		return sess, nil
+	} else if err != mongo.ErrNoDocuments {
+		return nil, errors.Wrap(err, "select")
+	}
+
+	return session.NewBaseSession(sid, s.encoder), nil
+
+}
+
+func (s mongoStore) Destroy(ctx context.Context, sid string) error {
+	_, err := s.db.Collection(s.collection).DeleteOne(ctx, bson.M{"key": sid})
+	if err != nil {
+		return errors.Wrap(err, "delete")
+	}
+	return nil
+}
+
+func (s mongoStore) Save(ctx context.Context, sess session.Session) error {
+	binary, err := sess.Encode()
+	if err != nil {
+		return errors.Wrap(err, "encode")
+	}
+
+	upsert := true
+	_, err = s.db.Collection(s.collection).
+		UpdateOne(ctx, bson.M{"key": sess.ID()}, bson.M{"$set": bson.M{
+			"key":        sess.ID(),
+			"data":       binary,
+			"expired_at": s.nowFunc().Add(s.lifetime).UTC(),
+		}}, &options.UpdateOptions{
+			Upsert: &upsert,
+		})
+	if err != nil {
+		return errors.Wrap(err, "upsert")
+	}
+	return nil
+}
+
+func (s mongoStore) GC(ctx context.Context) error {
+	_, err := s.db.Collection(s.collection).DeleteMany(ctx, bson.M{"expired_at": bson.M{"$lt": s.nowFunc().UTC()}})
+	if err != nil {
+		return errors.Wrap(err, "GC")
+	}
+	return nil
+}
+
+// Options keeps the settings to set up Redis client connection.
+type Options = options.ClientOptions
+
+// Config contains options for the Redis session store.
+type Config struct {
+	// For tests only
+	nowFunc func() time.Time
+	db      *mongo.Database
+
+	// Options is the settings to set up MongoDB client connection.
+	Options *Options
+	// DSN is the database source name to the MongoDB.
+	DSN string
```

Comment:
```suggestion
	// Database is the database name of the MongoDB.
	Database string
```

Based on my understanding, this field is purely used as a database name, not as a DSN.